### ISSUE TITLE
[tune]  BOHB algorithm; role of num_samples and custom early stopping

### ISSUE THREAD

**Issue by bask0** [ray-project/ray#5775]
### System information
- **Ray version**:
0.8.0.dev4

### Describe the problem

I have two questions regarding the [BOHB algorithm](https://ray.readthedocs.io/en/latest/tune-schedulers.html#hyperband-bohb):

* It is not clear to me how `tune.run(num_samples=...)` interacts with the BOBH algorithm (`TuneBOHB` and `HyperBandForBOHB`). When I set `num_samples=10` for example, only 10 trials are used in total. However, the documentation refers to [this post](https://homes.cs.washington.edu/~jamieson/hyperband.html), where I cannot find any parameter referring to a maximum number of trials / samples (i.e. `num_samples`). Here, I set `num_samples=80`, for example:

```
Round #0:
  Bracket(Max Size (n)=1, Milestone (r)=64, completed=100.0%): {TERMINATED: 64} 
  Bracket(Max Size (n)=2, Milestone (r)=76, completed=53.4%): {RUNNING: 2, TERMINATED: 14} 
```

* Is it possible to add custom early stopping (when the validation set performance drops a number of steps)? This would cause some trials to stop even if the BOHB algorithm would run more steps. Can the algorithm handle such "incomplete" runs? Example: BOHB wants to run a trial until epoch 100. After 80 epochs, the validation loss gets worse over several epochs. The trial is stopped through `tune.run(..., stop={'patience_counter': 3}`.

**Comment by richardliaw** [comment]
1. If you want the closest HyperBand setup to the blog, you need `num_samples= $\sum_{i} n_i$` (so 81 + 27 + 9 + 6 + 5 if eta = 3). 
2. IIUC trials can be killed and BOHB should do the bookkeeping correctly; though if you run into issues, please let me know.

**Comment by bask0** [comment]
Thank you very much, you helped me a lot! I'll let you know if 2. causes problems.

**Comment by richardliaw** [comment]
BTW, feel free to join our slack to chat faster (https://docs.google.com/forms/d/e/1FAIpQLSfAcoiLCHOguOm8e7Jnn-JJdZaCxPGjgVCvFijHB5PLaQLeig/viewform)
